core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala

```diff
@@ -160,11 +160,29 @@ case class SparkListenerBlockUpdated(blockUpdatedInfo: BlockUpdatedInfo) extends
  * Periodic updates from executors.
  * @param execId executor id
  * @param accumUpdates sequence of (taskId, stageId, stageAttemptId, accumUpdates)
+ * @param executorUpdates executor level metrics updates
  */
 @DeveloperApi
 case class SparkListenerExecutorMetricsUpdate(
     execId: String,
-    accumUpdates: Seq[(Long, Int, Int, Seq[AccumulableInfo])])
+    accumUpdates: Seq[(Long, Int, Int, Seq[AccumulableInfo])],
+    executorUpdates: Option[Array[Long]] = None)
+  extends SparkListenerEvent
+
+/**
+ * Peak metric values for the executor for the stage, written to the history log at stage
+ * completion.
+ * @param execId executor id
+ * @param stageId stage id
+ * @param stageAttemptId stage attempt
+ * @param executorMetrics executor level metrics, indexed by MetricGetter.values
+ */
+@DeveloperApi
+case class SparkListenerStageExecutorMetrics(
+    execId: String,
+    stageId: Int,
+    stageAttemptId: Int,
+    executorMetrics: Array[Long])
```

Comment:
We cannot expose an array of long's in a developer api (mutability).
In addition, we cannot have users needing to reference private spark api's to understand the meaning of it - particularly when the ordering can be subject to change in subsequent versions of spark.

Reply:
We can change back to using an ExecutorMetrics class in this case.

The plan was for any new metrics to be added to the end, so that there wouldn't be any change in ordering, and executorMetrics could be changed to immutable Seq[Long], but there would still be the issue of having to reference MetricGetter to find out how the metrics are indexed.

Reply:
sorry was out for a bit, catching up now -- 

so I think these are valid concerns, but I'd really like to have an approach which avoids all the getFooBar() methods -- its more error prone internally with all the places you've got to make sure you match things up exactly, and even for external consumers its nice to have an easy way to iterate over everything.  I think we can satisfy all concerns by making the field & constructor private and adding a `getMetricValue(metric: MetricGetter)`.  perhaps we should also rename `MetricGetter` -- externally its more like `ExecutorMetricType`.  Users shouldn't know anything about its ability to grab the value of the metric in a running executor.

I'm not totally stuck on this, but I'd certainly appreciate it -- I'm very partial to enums, but realize not everyone sees the beauty of them like I do :)

Reply:
Adding getMetricValue() would abstract away the array implementation. Should MetricGetter/ExecutorMetricType be public?

Reply:
+1 on enum's @squito !
The only concern would be evolving the enum's in a later release - changing enum could result in source incompatibility.

Reply:
re: compatability -- the json already uses names, so is not dependent on enum order.  we always require the same code for the values sent between the driver & executor, so that isn't a concern.  For the user accessing these values with `getMetricValue()` -- deleting an enum would be the same as deleting a field, so yeah it would break compatibility.  technically allowed for a "developerapi" but something we should avoid.  Adding enums should be OK.  If the enums are re-ordered in the spark code, but the user compiles against an older version ... I *think* it should be OK, as we'd look up the index in the actual spark code at runtime.

btw, I'm using "enum" loosely, I don't actually mean a java enum, I mean more the general concept, as its implemented in the current diff.  A fixed set of constants, with a helper to get all of them in order.  We could switch to using java enums -- I actually started that (https://github.com/apache/spark/pull/21221/commits/8b74ba8fff21b499e7cc9d93f9864831aa29773e), but changed it in the end.  honestly I forget why -- I think it was because I decided I wanted scala scoping constructs and more flexible code organization or something along those lines, and java's enum didn't really buy me much more.  The `executorMetrics` field here is basically an `EnumMap`, but it can actually do primitives.  That matters more for the internal messages, than here in the public spark listener api.

anyway, I think there *does* need to be some developer api which exposes all of the MetricGetter/ExecutorMetricType values.  I don't really care whether that is a java enum, or the home-grown version here.  I'm flexible on specifics, but my suggestion: an `ExecutorMetricType` enum that is a developerapi; make the SparkListener api expose the values as `executorMetrics: EnumMap<ExecutorMetricType, Long>`; internally, still use `Array[Long]`, and have `MetricGetter` contain the code which knows how to take an executorMetricType and determine its current value.  that would make the public api more self-documenting and clean, keep things efficient and clean internally, and also allow us to separate out the public list of metrics from the internal logic to determine current values, without having to play too many games with package-private access.

Reply:
This sounds like a good solution, with both a clean API, and also an efficient implementation that will be easier to add new metrics to. Thanks!

Reply:
errrr ... actually just realized EnumMap is always mutable.  If we want the public api to be totally immutable, back to private field & constructor, and a getter.  so might as well keep the field an `Array[Long]` since its private anyway.  IMO doesn't really matter if `ExecutorMetricType` is an enum or a sealed scala trait.